File: 剑指offer/51.数组中的逆序对.py
from typing import List


class Solution:
    def reversePairs(self, nums: List[int]) -> int:
        size = len(nums)
        if size < 2:
            return 0

        temp = [0 for _ in range(size)]
        return self.count_reverse_pairs(nums, 0, size - 1, temp)

    def count_reverse_pairs(self, nums, left, right, temp):
        if left == right:
            return 0

        mid = (left + right) >> 1
        left_pairs = self.count_reverse_pairs(nums, left, mid, temp)
        right_pairs = self.count_reverse_pairs(nums, mid + 1, right, temp)

        reverse_pairs = left_pairs + right_pairs

        if nums[mid] <= nums[mid + 1]:
            return reverse_pairs

        reverse_cross_pairs = self.merge_and_count(nums, left, mid, right,
                                                   temp)
        return reverse_pairs + reverse_cross_pairs

    def merge_and_count(self, nums, left, mid, right, temp):
        for i in range(left, right + 1):
            temp[i] = nums[i]

        i = left
        j = mid + 1
        res = 0
        for k in range(left, right + 1):
            if i > mid:
                nums[k] = temp[j]
                j += 1
            elif j > right:
                nums[k] = temp[i]
                i += 1
            elif temp[i] <= temp[j]:
                nums[k] = temp[i]
                i += 1
            else:
                nums[k] = temp[j]
                j += 1
                res += (mid - i + 1)
        return res

File: 剑指offer/51.数组中的逆序对.cpp
#include <vector>
using namespace std;
class Solution {
public:
    int mergeSort(vector<int>& nums, vector<int>& tmp, int left, int right)
    {
        if (left >= right) {
            return 0;
        }

        int mid = left + (right - left) / 2;
        int leftNum = mergeSort(nums, tmp, left, mid);
        int rightNum = mergeSort(nums, tmp, mid + 1, right);

        int pos = left;
        int l = left;
        int r = mid + 1;
        int crossNum = 0;

        while (l <= mid && r <= right) {
            if (nums[r] < nums[l]) {
                crossNum += (mid - l + 1);
                tmp[pos] = nums[r];
                r++;
            } else {
                tmp[pos] = nums[l];
                l++;
            }
            pos++;
        }

        while (l <= mid) {
            tmp[pos] = nums[l];
            l++;
            pos++;
        }

        while (r <= right) {
            crossNum += (mid - l + 1);
            tmp[pos] = nums[r];
            r++;
            pos++;
        }
        copy(tmp.begin() + left, tmp.begin() + right + 1, nums.begin() + left);

        return leftNum + rightNum + crossNum;
    }

    int reversePairs(vector<int>& nums)
    {
        int n = nums.size();
        vector<int> tmp(n);
        return mergeSort(nums, tmp, 0, n - 1);
    }
};